config/v1alpha1/types_backup.go

```diff
@@ -0,0 +1,147 @@
+	// +kubebuilder:validation:Required
+	EtcdBackupSpec EtcdBackupSpec `json:"etcdBackupSpec"`
+}
+
+type BackupStatus struct {
+}
+
+// EtcdBackupSpec provides configuration for automated etcd backups to the cluster-etcd-operator
+type EtcdBackupSpec struct {
+
+	// Schedule defines the recurring backup schedule in Cron format
+	// every 2 hours: 0 */2 * * *
+	// every day at 3am: 0 3 * * *
+	// Setting to an empty string "" means disabling scheduled backups
+	// Default: ""
+	// TODO: Define CEL validation for the cron format
+	// See if the upstream CronJob has CEL validation already defined somewhere.
+	// TODO: Validation to disallow unrealistic schedules (e.g */2 * * * * every 2 mins)
+	// +kubebuilder:validation:Optional
+	// +optional
+	Schedule string `json:"schedule"`
+
+	// The time zone name for the given schedule, see https://en.wikipedia.org/wiki/List_of_tz_database_time_zones.
+	// If not specified, this will default to the time zone of the kube-controller-manager process.
+	// See https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#time-zones
+	// +kubebuilder:validation:Optional
+	// +optional
+	// TODO: CEL validation for timezone format
+	TimeZone string `json:"timeZone,omitempty"`
+
+	// RetentionPolicy defines the retention policy for retaining and deleting existing backups.
+	// +kubebuilder:validation:Optional
+	// +optional
+	RetentionPolicy RetentionPolicy `json:"retentionPolicy"`
+
+	// PVCName specifies the name of the PersistentVolumeClaim which binds a PersistentVolume where the
+	// etcd backup files would be saved
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	PVCName string `json:"pvcName"`
+}
+
+// RetentionType is the enumeration of valid retention policy types
+// +enum
+// +kubebuilder:validation:Enum:="RetentionCount";"RetentionSize"
+type RetentionType string
+
+const (
+	// RetentionTypeCount sets the retention policy based on the count or number of backup files saved
+	RetentionTypeCount RetentionType = "RetentionCount"
+	// RetentionTypeSize sets the retention policy based on the total size of the backup files saved
+	RetentionTypeSize RetentionType = "RetentionSize"
+)
+
+// RetentionPolicy defines the retention policy for retaining and deleting existing backups.
+// This struct is a discriminated union that allows users to select the type of retention policy from the supported types.
+// +union
+type RetentionPolicy struct {
+	// RetentionType sets the type of retention policy. The currently supported and valid values are "retentionCount"
+	// Currently, the only valid policies are retention by count (RetentionCount) and by size (RetentionSize). More policies or types may be added in the future.
+	// +unionDiscriminator
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=RetentionCount;RetentionSize
+	// TODO: How to provide a "NoRetentionPolicy" option as a default? Should that be another enum
```

Comment:
just allow empty

Reply:
Just to clarify, since the discriminant is always required, that can't be empty, so I'm assuming you mean just allowing empty values for the union members right?
https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions
> - The discriminant should be required
> - Empty union members (discriminant values without a paired union member) are also permitted

So we would interpret the following configs as "NoRetentionPolicy":
```yaml
spec:
  etcd:
    retentionPolicy:
      retentionType: RetentionNumber
      # No member specified
```
```yaml
spec:
  etcd:
    retentionPolicy:
      retentionType: RetentionSize
      # No member specified
```

The `retentionPolicy` field is also optional itself so leaving that empty would also be valid:
```yaml
spec:
  etcd:
    schedule: * 2 * * *
    # No retentionPolicy specified
```
Is that what you meant?

Reply:
Ah nevermind, just saw [your comment](https://github.com/openshift/api/pull/1482#discussion_r1245699431) below for allowing an empty string for the discriminant.